File: plugins/weather/weather.30m.go
//go:build ignore

package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
)

func main() {
	apiKey := os.Getenv("WEATHER_API_KEY")
	city := os.Getenv("WEATHER_CITY")
	if city == "" {
		city = "London"
	}

	if apiKey == "" {
		fmt.Println("🌡️ No API Key")
		fmt.Println("---")
		fmt.Println("Set WEATHER_API_KEY")
		return
	}

	url := fmt.Sprintf("https://api.openweathermap.org/data/2.5/weather?q=%s&appid=%s&units=metric", city, apiKey)
	resp, err := http.Get(url)
	if err != nil {
		fmt.Println("🌡️ Error")
		return
	}
	defer resp.Body.Close()

	var data map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		fmt.Println("🌡️ Parse Error")
		return
	}

	temp := "--"
	desc := ""
	if main, ok := data["main"].(map[string]interface{}); ok {
		if t, ok := main["temp"].(float64); ok {
			temp = fmt.Sprintf("%.1f", t)
		}
	}
	if weather, ok := data["weather"].([]interface{}); ok && len(weather) > 0 {
		if w, ok := weather[0].(map[string]interface{}); ok {
			desc, _ = w["description"].(string)
		}
	}

	fmt.Printf("🌡️ %s°C\n", temp)
	fmt.Println("---")
	fmt.Printf("Location: %s\n", city)
	fmt.Printf("Temperature: %s°C\n", temp)
	fmt.Printf("Condition: %s\n", desc)
	fmt.Println("---")
	fmt.Println("Refresh | refresh=true")
}

File: plugins/weather/weather.30m.rs
//! Weather Plugin - Uses Crossbar API for HTTP requests

use std::env;
use std::process::Command;
use serde_json::Value;

fn crossbar(args: &[&str]) -> Option<String> {
    let output = Command::new("crossbar")
        .args(args)
        .output()
        .ok()?;
    
    if output.status.success() {
        Some(String::from_utf8_lossy(&output.stdout).trim().to_string())
    } else {
        None
    }
}

fn main() {
    let api_key = env::var("WEATHER_API_KEY").unwrap_or_default();
    let city = env::var("WEATHER_CITY").unwrap_or_else(|_| "London".to_string());

    if api_key.is_empty() {
        println!("🌡️ No API Key");
        println!("---");
        println!("Set WEATHER_API_KEY in configuration");
        return;
    }

    let url = format!(
        "https://api.openweathermap.org/data/2.5/weather?q={}&appid={}&units=metric",
        city, api_key
    );
    
    let response = crossbar(&["--web", &url, "--json"]);

    if response.is_none() {
        println!("🌡️ Error");
        println!("---");
        println!("Failed to fetch weather data");
        return;
    }

    let response = response.unwrap();
    match serde_json::from_str::<Value>(&response) {
        Ok(data) => {
            let temp = data["main"]["temp"].as_f64()
                .map(|t| format!("{:.1}", t))
                .unwrap_or_else(|| "--".to_string());
            let desc = data["weather"][0]["description"]
                .as_str()
                .unwrap_or("");

            println!("🌡️ {}°C", temp);
            println!("---");
            println!("Location: {}", city);
            println!("Temperature: {}°C", temp);
            println!("Condition: {}", desc);
        }
        Err(_) => println!("🌡️ Parse Error"),
    }

    println!("---");
    println!("Refresh | refresh=true");
}
